### ISSUE TITLE
[Docker] add a `onbuild` image to Wechaty

### ISSUE THREAD

**Issue by zixia** [Chatie/wechaty#147]
[![Bountysource](https://api.bountysource.com/badge/issue?issue_id=40206801)](https://www.bountysource.com/issues/40206801-docker-add-a-onbuild-image-to-wechaty?utm_source=40206801&utm_medium=shield&utm_campaign=ISSUE_BADGE)

Gist discussion with @imWildCat at [here](https://gist.github.com/imWildCat/4eb264ed33e69c70cb391bc68aa48810)

**Comment by zixia** [comment]
Good news: I just made a Wechaty **onbuild** image this night. 

Basically, you can build your own Wechaty Bot docker image very easy by following 3 steps:

1. Make sure your project directory work well by `npm install && npm start`
1. Create a `Dockerfile` at the project top directory with the content: `FROM: zixia/wechaty:onbuild`
1. Run `docker build -t wechaty-bot .`

Please report the issue if it does not work as expected.

Please see more detail on our wiki at:
<https://github.com/Chatie/wechaty/wiki/Docker#onbuild>

**Comment by imWildCat** [comment]
![snipaste_20170424_155309](https://cloud.githubusercontent.com/assets/2396817/25343161/231e5ccc-2906-11e7-85bd-44945bf4767c.png)

It works very fine! Thanks a lot!

```dockerfile
FROM zixia/wechaty:onbuild
ENV NODE_ENV production
RUN mkdir -p /bot/data # only for the store of *.wechaty.json
ADD ./ /bot

CMD ["npm", "start"]
```

**Comment by zixia** [comment]
@imWildCat indeed, your project should work by using this Dockerfile:

```dockerfile
FROM zixia/wechaty:onbuild
```

That's all. :)

**Comment by imWildCat** [comment]
Issue report:

> The installation directory of the custom dependencies in the `package.json` of my project are in `/node_modules` instead of `/bot/node_modules`.

As a result, if the execution will fail because the programme cannot find my dependencies.

**Comment by zixia** [comment]
I'll have a look later.

However, use the root directory to install node_modules is by design.

The modules that installed into /node_modules should be ok because Node.js will search parent directory for node_module.

> If the module identifier passed to require() is not a core module, and does not begin with '/', '../', or './', then Node.js starts at the parent directory of the current module, and adds /node_modules, and attempts to load the module from that location. Node will not append node_modules to a path already ending in node_modules.

> If it is not found there, then it moves to the parent directory, and so on, until the root of the file system is reached.

See: <https://nodejs.org/api/modules.html#modules_loading_from_node_modules_folders>

**Comment by imWildCat** [comment]
@zixia I have looked at this issue in 'depth'. As for this project https://github.com/mukaiu/bot-docker-test, the module 'serve-favicon' seemed not to be installed in `/node_modules` correctly. Instead, it is in these directories:
```shell
/root/.npm/serve-favicon
/root/.npm/registry.npmjs.org/serve-favicon
/tmp/npm-5-06ef38e2/registry.npmjs.org/serve-favicon
/tmp/npm-7-bc212f86/registry.npmjs.org/serve-favicon
```

Way for recurrence (in the project mentioned above):
```shell
bash-4.3# ./bin/www
module.js:472
    throw err;
    ^

Error: Cannot find module 'serve-favicon'
    at Function.Module._resolveFilename (module.js:470:15)
    at Function.Module._load (module.js:418:25)
    at Module.require (module.js:498:17)
    at require (internal/module.js:20:19)
    at Object.<anonymous> (/bot/app.js:3:15)
    at Module._compile (module.js:571:32)
    at Object.Module._extensions..js (module.js:580:10)
    at Module.load (module.js:488:32)
    at tryModuleLoad (module.js:447:12)
    at Function.Module._load (module.js:439:3)
bash-4.3# find / -name serve-favicon
/root/.npm/serve-favicon
/root/.npm/registry.npmjs.org/serve-favicon
/tmp/npm-5-06ef38e2/registry.npmjs.org/serve-favicon
/tmp/npm-7-bc212f86/registry.npmjs.org/serve-favicon
```

**Comment by zixia** [comment]
Yes, I'm looking on this too.

It's very strange that the `npm install` command seems not working right.

**Comment by zixia** [comment]
@imWildCat @mukaiu Hi all, after did some tests, I found that it's related to the directory name: `/bot`.

`/bot` is the VOLUME of Wechaty Docker Image, we all know about it because we use `--volume="$(pwd)":/bot` to map local directory into Docker.

The previous version of our onbuild Docker image use `/bot` as WORKDIR, it works fine unless we want to do an `npm install` to it. It seems the directory will not store the `'node_modules` directory for us for some reason, which I still have no idea of.

However, if we change the WORKDIR to another name, ie `/app`, then the `npm install` works without any problem.

Today I had changed toe WORKDIR to `/app` and everything looks ok now. Please check out the latest version(v0.7.75 or above) of Wechaty Docker and have a try, and let me hear back from you.

Thanks!

**Comment by zixia** [comment]
@imWildCat Let's talk about this strange `VOLUME/WORKDIR` bug at #436

**Comment by ditunes** [comment]
still has problem ,this is my error info 
```
wechat-bot@1.0.0 start /app
> tsc && node ./dist/index.js

03:01:22 INFO Wechaty v0.7.71 initializing...
03:01:23 WARN PuppetWebBrowserDriver valid() getSession() catch() rejected: Server terminated early with status 1
03:01:23 WARN PuppetWebBrowserDriver valid() driver.getSession() exception: Server terminated early with status 1
03:01:23 WARN PuppetWebBrowserDriver initChromeDriver() got invalid driver
03:01:23 WARN PuppetWebBrowserDriver initChromeDriver() with retry: 1
(node:33) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 1): Error: Server terminated early with status 1
(node:33) DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
03:01:23 WARN PuppetWebBrowserDriver valid() getSession() catch() rejected: Server terminated early with status 1
03:01:23 WARN PuppetWebBrowserDriver valid() driver.getSession() exception: Server terminated early with status 1
03:01:23 WARN PuppetWebBrowserDriver initChromeDriver() got invalid driver
03:01:23 WARN PuppetWebBrowserDriver initChromeDriver() with retry: 2
(node:33) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 2): Error: Server terminated early with status 1
```

my dockerfile is 
```

FROM zixia/wechaty:onbuild
#WORKDIR /bot
RUN mkdir -p /bot/data
ADD ./ /bot
CMD [ "npm","start" ]


```

**Comment by zixia** [comment]
Please try to change your Dockerfile as this:

```
FROM zixia/wechaty:onbuild
```

Then let me know if it works.

**Comment by imWildCat** [comment]
@ditunes With the latest version of 'onbuild', your project was running successfully.

**Comment by zixia** [comment]
Hope everything goes well with the onbuild image.

I'll close this issue, and please create a new issue if there's any problem with the docker image.

**Comment by zixia** [comment]
@ditunes Hello, would you like to write a short article to introduce how to use Wechaty:onbuild Docker Image to build a ChatBot docker image by yourself? 

Because this could help other developers to save their time to use it.

**Comment by imWildCat** [comment]
I am currently learning TensorFlow. If everything is fine, I am considering to write an article about chatbot and TensorFlow. If possible, using TensorFlow (cuda) within Docker can be mentioned. However, it is also too specific rather than general. And it will take me a long time.

**Comment by zixia** [comment]
@imWildCat Awesome!

Would you like to use Wechaty as the ChatBot example and post it to blog.chatie.io also?

**Comment by imWildCat** [comment]
@zixia Of course. That's also the direction for my application.
But please do not push me =。=
It should be finished before July.

**Comment by zixia** [comment]
Great! I had just created an issue on the blog for a memo.

Take your time, no rush, be cool. :)

**Comment by zixia** [comment]
@ditunes Hello, would you like to share your Docker Onbuild experience for others? 

I'm looking forward to your reply.

Thanks!

**Comment by MaxLee1994** [comment]
It seems that I have encountered the same error with @ditunes 

Here's my error info

    Starting Wechaty v0.8.14 with Node.js v7.10.0 ...

    > wechat-watch-bot@0.1.0 start /app
    > ts-node src/index.ts

    03:35:01 INFO Wechaty v0.7.95 initializing...
    03:35:01 WARN PuppetWebBrowserDriver valid() getSession() catch() rejected: Server terminated early with status 1
    03:35:01 WARN PuppetWebBrowserDriver valid() driver.getSession() exception: Server terminated early with status 1
    03:35:01 WARN PuppetWebBrowserDriver initChromeDriver() got invalid driver
    03:35:01 WARN PuppetWebBrowserDriver initChromeDriver() with retry: 1
    (node:38) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 1): Error: Server terminated early with status 1
    (node:38) DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
    03:35:01 WARN PuppetWebBrowserDriver valid() getSession() catch() rejected: Server terminated early with status 1
    03:35:01 WARN PuppetWebBrowserDriver valid() driver.getSession() exception: Server terminated early with status 1
    03:35:01 WARN PuppetWebBrowserDriver initChromeDriver() got invalid driver
    03:35:01 WARN PuppetWebBrowserDriver initChromeDriver() with retry: 2
    (node:38) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 2): Error: Server terminated early with status 1
    03:35:01 WARN PuppetWebBrowserDriver valid() getSession() catch() rejected: Server terminated early with status 1
    03:35:01 WARN PuppetWebBrowserDriver valid() driver.getSession() exception: Server terminated early with status 1
    03:35:01 WARN PuppetWebBrowserDriver initChromeDriver() got invalid driver
    03:35:01 WARN PuppetWebBrowserDriver initChromeDriver() with retry: 3
    (node:38) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 3): Error: Server terminated early with status 1
    03:35:01 WARN PuppetWebBrowserDriver valid() getSession() catch() rejected: Server terminated early with status 1
    03:35:01 WARN PuppetWebBrowserDriver valid() driver.getSession() exception: Server terminated early with status 1
    03:35:01 WARN PuppetWebBrowserDriver initChromeDriver() got invalid driver
    03:35:01 WARN PuppetWebBrowserDriver initChromeDriver() not valid after retry: 4 times: Error: initChromeDriver() got invalid driver
        at BrowserDriver.<anonymous> (/app/node_modules/wechaty/src/puppet-web/browser-driver.ts:146:21)
        at Generator.next (<anonymous>)
        at fulfilled (/app/node_modules/wechaty/dist/src/puppet-web/browser-driver.js:4:58)
        at process._tickDomainCallback (internal/process/next_tick.js:135:7)
    03:35:01 ERR PuppetWebBrowser init() exception: initChromeDriver() got invalid driver
    03:35:01 ERR PuppetWebBrowser quit() this.driver.quit() exception Server terminated early with status 1
    (node:38) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 4): Error: Server terminated early with status 1
    03:35:01 ERR PuppetWeb initBrowser() exception: initChromeDriver() got invalid driver
    03:35:01 ERR PuppetWeb init() exception: Error: initChromeDriver() got invalid driver
        at BrowserDriver.<anonymous> (/app/node_modules/wechaty/src/puppet-web/browser-driver.ts:146:21)
        at Generator.next (<anonymous>)
        at fulfilled (/app/node_modules/wechaty/dist/src/puppet-web/browser-driver.js:4:58)
        at process._tickDomainCallback (internal/process/next_tick.js:135:7)
    (node:38) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 6): TypeError: Cannot read property 'quit' of undefined
    03:35:08 WARN PuppetWebBrowserDriver valid() driver.getSession() timeout(halt?)
    03:35:08 WARN PuppetWebBrowserDriver valid() driver.getSession() timeout(halt?)
    03:35:08 WARN PuppetWebBrowserDriver valid() driver.getSession() timeout(halt?)
    03:35:08 WARN PuppetWebBrowserDriver valid() driver.getSession() timeout(halt?)
    03:37:01 WARN PuppetWeb quit() Promise() timeout
    03:37:01 ERR PuppetWeb quit() exception: quit() Promise() timeout
    03:37:01 ERR Wechaty init() exception: quit() Promise() timeout
    (node:38) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 7): Error: quit() Promise() timeout

My Dockerfile's just one line `FROM zixia/wechaty:onbuild`

**Comment by MaxLee1994** [comment]
Turns out that I have local dependency of `wechaty` in my `package.json`.
Found this solution in #281 

Perhaps it's best to document this in the wiki.
Fantastic project btw🎉

**Comment by zixia** [comment]
@MaxLi1994 Thanks for pointing out this problem again.

Do not have local dependency is good practice.

But it should not break your code. There must be some bug inside somewhere.

**Comment by MaxLi1994** [comment]
@zixia The bug didn't show up after I had removed local dependency of `wechaty`. 

I noticed that my local version of `wechaty` was 0.7.x, perhaps it's the old version that caused the bug.

**Comment by zixia** [comment]
No news is good news. 

Issue closed.
